File: src/simulation/infrastructure.js
import { recordEvent } from "./journal.js";

const MONTHLY_PER_PERSON = {
  timber: 0.001,
  clay: 0.0007,
  stone: 0.0005,
  tools: 0.00004
};

export function infrastructureMonthlyNeed(city, population, resourceId) {
  return population * (MONTHLY_PER_PERSON[resourceId] ?? 0) *
    (1.15 - city.infrastructure.housingCondition * 0.25);
}

function clamp01(value) {
  return Math.max(0, Math.min(1, value));
}

export function advanceInfrastructure(world, telemetry) {
  if (world.day === 0 || world.day % 30 !== 0) return;
  for (const cityId of Object.keys(world.cities).sort()) {
    const city = world.cities[cityId];
    const population = world.spatial.nodes[city.spatialNodeId].aggregate.population;
    const ratios = {};
    for (const resourceId of Object.keys(MONTHLY_PER_PERSON).sort()) {
      const needed = infrastructureMonthlyNeed(city, population, resourceId);
      const consumed = Math.min(needed, city.stocks[resourceId]);
      city.stocks[resourceId] = Math.round(Math.max(0, city.stocks[resourceId] - consumed) * 1_000_000) / 1_000_000;
      ratios[resourceId] = needed > 0 ? consumed / needed : 1;
      telemetry.infrastructureConsumptionByResource[resourceId] =
        Math.round(((telemetry.infrastructureConsumptionByResource[resourceId] ?? 0) + consumed) * 1_000_000) / 1_000_000;
    }
    const housingSupply = Math.min(ratios.timber, ratios.clay);
    const roadSupply = Math.min(ratios.stone, ratios.tools);
    const previousHousing = city.infrastructure.housingCondition;
    city.infrastructure.housingCondition = clamp01(previousHousing + (housingSupply - 0.9) * 0.012);
    city.infrastructure.roadCondition = clamp01(city.infrastructure.roadCondition + (roadSupply - 0.88) * 0.009);
    city.infrastructure.sanitation = clamp01(city.infrastructure.sanitation +
      (Math.min(ratios.clay, city.shortage.active ? 0.25 : 1) - 0.9) * 0.006);

    if (previousHousing >= 0.5 && city.infrastructure.housingCondition < 0.5) {
      const causeIds = Object.entries(ratios)
        .filter(([, ratio]) => ratio < 0.5)
        .map(([resourceId]) => city.resourceSignals[resourceId])
        .filter(Boolean);
      recordEvent(world, {
        type: "infrastructure_degraded",
        subjectId: cityId,
        causeIds,
        details: { cityId, component: "housing", condition: city.infrastructure.housingCondition, supplyRatios: ratios }
      });
    }
  }

  for (const route of [...world.routes].sort((left, right) => left.id.localeCompare(right.id))) {
    const endpointCondition = (world.cities[route.a].infrastructure.roadCondition +
      world.cities[route.b].infrastructure.roadCondition) / 2;
    const traffic = world.journal
      .filter((event) => event.day >= world.day - 30 && event.type === "shipment_dispatched" &&
        event.details.routeIds.includes(route.id))
      .reduce((sum, event) => sum + event.details.amount, 0);
    const utilization = traffic / Math.max(1, route.baseDailyCapacity * 30);
    const previousCondition = route.condition;
    route.condition = clamp01(route.condition + (endpointCondition - route.condition) * 0.035 -
      Math.min(0.006, utilization * 0.003));
    route.travelDays = Math.max(1, Math.round(route.baseTravelDays * (1.32 - route.condition * 0.32)));
    route.dailyCapacity = Math.round(route.baseDailyCapacity * (0.55 + route.condition * 0.45) * 1_000_000) / 1_000_000;
    if (previousCondition >= 0.4 && route.condition < 0.4) {
      recordEvent(world, {
        type: "infrastructure_degraded",
        subjectId: route.id,
        details: { component: "route", routeId: route.id, condition: route.condition, utilization }
      });
    }
  }
}

File: src/WorldGen.Core/Simulation/Infrastructure.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace WorldGen.Core.Simulation;

public static class Infrastructure
{
    public static void Advance(WorldState world, DailyTelemetry telemetry)
    {
        if (world.Day == 0 || world.Day % 30 != 0) return;
        foreach (var cityId in world.Cities.Keys.Order(StringComparer.Ordinal))
        {
            var city = world.Cities[cityId]; var population = world.Spatial.Nodes[city.SpatialNodeId].Aggregate.Population;
            var ratios = new Dictionary<string, double>(StringComparer.Ordinal);
            foreach (var resourceId in NeedsAndDemand.InfrastructureResourceIds)
            {
                var needed = NeedsAndDemand.InfrastructureMonthlyNeed(city, population, resourceId);
                var consumed = Math.Min(needed, city.Stocks[resourceId]);
                city.Stocks[resourceId] = SimulationMath.Quantize(Math.Max(0, city.Stocks[resourceId] - consumed));
                ratios[resourceId] = needed > 0 ? consumed / needed : 1;
                telemetry.InfrastructureConsumptionByResource[resourceId] = SimulationMath.Quantize(
                    telemetry.InfrastructureConsumptionByResource.GetValueOrDefault(resourceId) + consumed);
            }
            var housingSupply = Math.Min(ratios["timber"], ratios["clay"]);
            var roadSupply = Math.Min(ratios["stone"], ratios["tools"]);
            var previous = city.Infrastructure.HousingCondition;
            city.Infrastructure.HousingCondition = SimulationMath.Clamp(previous + (housingSupply - 0.9) * 0.012, 0, 1);
            city.Infrastructure.RoadCondition = SimulationMath.Clamp(city.Infrastructure.RoadCondition + (roadSupply - 0.88) * 0.009, 0, 1);
            city.Infrastructure.Sanitation = SimulationMath.Clamp(city.Infrastructure.Sanitation +
                (Math.Min(ratios["clay"], city.Shortage.Active ? 0.25 : 1) - 0.9) * 0.006, 0, 1);
            if (previous >= 0.5 && city.Infrastructure.HousingCondition < 0.5)
            {
                var causes = ratios.Where(pair => pair.Value < 0.5)
                    .Select(pair => city.ResourceSignals.GetValueOrDefault(pair.Key)).Where(id => id is not null);
                Journal.Record(world, "infrastructure_degraded", cityId, causes,
                    new JsonObject { ["cityId"] = cityId, ["component"] = "housing", ["condition"] = city.Infrastructure.HousingCondition,
                        ["supplyRatios"] = JsonSerializer.SerializeToNode(ratios) });
            }
        }

        foreach (var route in world.Routes.OrderBy(route => route.Id, StringComparer.Ordinal))
        {
            var endpoint = (world.Cities[route.A].Infrastructure.RoadCondition + world.Cities[route.B].Infrastructure.RoadCondition) / 2;
            var traffic = world.Journal.Where(evt => evt.Day >= world.Day - 30 && evt.Type == "shipment_dispatched" &&
                evt.Details["routeIds"]!.AsArray().Any(node => node!.GetValue<string>() == route.Id))
                .Sum(evt => evt.Details["amount"]!.GetValue<double>());
            var utilization = traffic / Math.Max(1, route.BaseDailyCapacity * 30);
            var previous = route.Condition;
            route.Condition = SimulationMath.Clamp(route.Condition + (endpoint - route.Condition) * 0.035 - Math.Min(0.006, utilization * 0.003), 0, 1);
            route.TravelDays = Math.Max(1, (int)SimulationMath.Round(route.BaseTravelDays * (1.32 - route.Condition * 0.32), 1));
            route.DailyCapacity = SimulationMath.Quantize(route.BaseDailyCapacity * (0.55 + route.Condition * 0.45));
            if (previous >= 0.4 && route.Condition < 0.4)
                Journal.Record(world, "infrastructure_degraded", route.Id, details:
                    new JsonObject { ["component"] = "route", ["routeId"] = route.Id, ["condition"] = route.Condition, ["utilization"] = utilization });
        }
    }
}
